lib/index.ts

```diff
@@ -78,18 +76,142 @@ export async function applyRemediationToManifests(
+    const trimmedText = requirementText.trim();
+
+    // Quick returns for cases we cannot remediate
+    // - Empty line i.e. ''
+    // - 'editable' packages i.e. '-e git://git.myproject.org/MyProject.git#egg=MyProject'
+    // - Comments i.e. # This is a comment
+    // - Local files i.e. file:../../lib/project#egg=MyProject
+    if (
+      requirementText === '' ||
+      trimmedText.startsWith('-e') ||
+      trimmedText.startsWith('#') ||
+      trimmedText.startsWith('file:')
+    ) {
+      return requirement;
+    }
+
+    // Regex to match against a Python package specifier. Any invalid lines (or
+    // lines we can't handle) should have been returned this point.
+    const regex = /([A-Z0-9]*)(===|==|>=|<=|>|<|~=)(\d\.?\d?\.?\d?)(.*)/i;
+    const result = regex.exec(requirementText);
+
+    if (result !== null) {
+      requirement.name = result[1];
+      requirement.versionComparator = result[2] as VersionComparator;
+      requirement.version = result[3];
+      requirement.extras = result[4];
+    }
+
+    return requirement;
+  });
+}
+
+// TS is not capable of determining when Array.filter has removed undefined
+// values without a manual Type Guard, so thats what this does
+function isDefined<T>(t: T | undefined): t is T {
+  return typeof t !== 'undefined';
+}
+
+function applyUpgrades(manifests: ManifestFiles, upgrades: DependencyUpdates) {
+  const requirementsFileName = Object.keys(manifests).find(
+    (fn) => path.basename(fn) === 'requirements.txt'
+  );
+
+  // If there is no usable manifest, return
+  if (typeof requirementsFileName === 'undefined') {
+    return manifests;
+  }
+
+  const requirementsFile = manifests[requirementsFileName];
+
+  const requirements = parseRequirementsFile(requirementsFile);
+
+  // This is a bit of a hack, but an easy one to follow. If a file ends with a
+  // new line, ensure we keep it this way. Don't hijack customers formatting.
+  let endsWithNewLine = false;
+  if (requirements[requirements.length - 1].originalText === '\n') {
+    endsWithNewLine = true;
+  }
+
+  const topLevelDeps = requirements
+    .map(({ name }) => name && name.toLowerCase())
+    .filter(isDefined);
+
+  const updatedRequirements: string[] = requirements.map(
+    ({ name, versionComparator, version, originalText, extras }) => {
+      // Defensive patching; if any of these are undefined, return
+      if (
+        typeof name === 'undefined' ||
+        typeof versionComparator === 'undefined' ||
+        typeof version === 'undefined'
+      ) {
+        return originalText;
+      }
+
+      // Check if we have an upgrade; if we do, replace the version string with
+      // the upgrade, but keep the rest of the content
+      const upgrade = upgrades[`${name.toLowerCase()}@${version}`];
 
-  applyUpgrades(manifests, upgrades, topLevelDeps);
+      if (!upgrade) {
+        return originalText;
+      }
+
+      const newVersion = upgrade.upgradeTo.split('@')[1];
```

Comment:
are we sure the package name will never have a '@' in the name?

Reply:
I believe so. See

https://pypi.org/search/?q=%22s%22&o= vs https://pypi.org/search/?q=%22%40%22&o=